eclipse-theia/theia
# vsx: support buildtime 'extensionPack' handling

#### What it does


Fixes: #9224 

The following commit adds improved support for `extensionPack` plugins defined as builtins. Previously, `extensionPacks` could be referenced as builtins, but they would later be resolved when the application is first started. For large packs (such as vscode builtins) this would result in a long startup time, where extensionPacks would need to be resolved, and then downloaded. The following commit fixes that problem by allowing `extensionPack` to be resolved and fetched at buildtime rather than
runtime.

In order to implement the improvement, a new dev-package was created (`@theia/ovsx-client`) which includes the common logic for the `cli` and `@theia/vsx-registry` (previously part of the `@theia/vsx-registry`) extension. This logic which interacts with the registry (fetch, search, download, determine compatibility) is now used by the `cli` (buildtime), and `vsx-registry` (runtime). The `@theia/vsx-registry` extension was refactored and cleaned up as a consequence.

The idea when resolving builtin extensions at buildtime is the following:
- download extensions like today (master).
- determine if any `extensionPack`(s) exist.
- fetch individual extension ids.
- download latest compatible versions of all individual extension ids.

#### How to test


`download-plugins cli`:

- verify that downloading plugins works like today, plugins are fetched and unpacked if necessary (based on flag).
- verify that if an extension pack is used as a builtin [example](https://github.com/vince-fugnitto/vscode-builtin-extensionPack/releases/download/0.0.1/builtin-extension-pack-1.45.1.vsix) it is downloaded, and for each `id` present in the pack the corresponding extension is fetched at its latest compatible version (based on application api level - this can be updated with a flag).
- verify that a mix of standard extensions and extension-pack work correctly - if an extension is present in both (listed normally and in the pack), the pack will not re-fetch the extension at a different version.

`vsx-registry`:

- verify that changes to vsx-registry still work correctly:
  - searching extensions with the extensions-view
  - installing/uninstalling extensions with the extensions-view
  - listing installed extensions, builtin extensions

#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>
Co-authored-by: marechal-p <paul.marechal@ericsson.com>
